sarif/sarif_file_utils.py

```diff
@@ -100,6 +100,103 @@ def read_result_location(result) -> Tuple[str, str]:
     return (file_path, line_number)
 
 
+def read_result_rule(result, run) -> Tuple[Union[dict, None], int]:
+    """
+    Return's the corresponding rule object for the specified result, plus its index
+    in the rules array. Follows the rules at
+    https://docs.oasis-open.org/sarif/sarif/v2.1.0/sarif-v2.1.0.html#_Toc141790895
+    """
+    ruleIndex = result.get("ruleIndex", None)
+    ruleId = result.get("ruleId", None)
+    rule = result.get("rule", None)
+
+    if not ruleIndex and rule:
+        ruleIndex = rule.get("index")
+
+    if not ruleId and rule:
+        ruleId = rule.get("id")
+
+    if not ruleIndex and not ruleId:
+        return (None, -1)
+
+    rules = run.get("tool", {}).get("driver", {}).get("rules", [])
+
+    if ruleIndex:
+        ruleIndex = int(ruleIndex)
+        return (rules[ruleIndex], ruleIndex)
+
+    for i, rule in enumerate(rules):
+        if rule.get("id") == ruleId:
+            return (rule, i)
+
+    return (None, -1)
+
+
+def read_result_invocation(result, run):
+    """
+    Extract the invocation metadata for the result, following the rules at
+    https://docs.oasis-open.org/sarif/sarif/v2.1.0/sarif-v2.1.0.html#_Toc141790917
+    """
+    invocationIndex = result.get("provenance", {}).get("invocationIndex", None)
+    if invocationIndex is None:
+        return None
+    
+    invocationIndex = int(invocationIndex)
+    if invocationIndex < 0:
+        return None
+
+    invocations = run.get("invocations")
+    return invocations[invocationIndex]
+
+
+def read_result_severity(result, run = {}) -> Literal["none", "note", "warning", "error"]:
+    """
+    Extract the severity level from the result following the rules at
+    https://docs.oasis-open.org/sarif/sarif/v2.1.0/sarif-v2.1.0.html#_Toc141790898
+    """
+    severity = result.get("level")
+    if severity:
+        return severity
+
+    # If kind has any value other than "fail", then if level is absent,
+    # it SHALL default to "none"
+    kind = result.get("kind", "fail")
+    if kind and kind != "fail":
+        return "none"
+
+    # If kind has the value "fail" and level is absent, then...
+    rule, ruleIndex = read_result_rule(result, run)
+    if rule:
+        # Honor the invocation's configuration override if present...
+        invocation = read_result_invocation(result, run)
+        if invocation:
+            ruleConfigurationOverrides = invocation.get("ruleConfigurationOverrides", [])
+            override = next(
+                (
+                    override
+                    for override in ruleConfigurationOverrides
+                    if override.get("descriptor").get("id") == rule.get("id") or
```

Comment:
`get("descriptor")` doesn't provide any safety against missing descriptors here,  maybe should be `get("descriptor", {})`?